### PR TITLE
Bounding box plugin update to TF2

### PR DESCRIPTION
**Pull request by dlfonseka**
This PR updates the ravenML-train bbox plugin `rmltraintfbbox` to use Tensorflow2 and the new TF2 updates to the Object Detection API. The old TF1 bbox plugin is preserved as `rmltraintfbboxlegacy`. 

Notables changes to the plugin:

- uses new 'Keras-like' detection models
- training loop less obfuscated
- restoring from frozen graph replaced with `SavedModel` restore
- evaluation is optional
- TFRecord data no longer copied to artifact directory

### CODE REVIEW

**Comment by jacobDeutsch10** [comment]
I have trained a few models using the updated bbox plugin, and it works well. LGTM!

File: rmltraintfbbox/rmltraintfbbox/utils/pipeline_templates/center_net_resnet101_v1_fpn_adam.config [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+  fine_tune_checkpoint: "<replace_arch_path>checkpoint/ckpt-0"
+  fine_tune_checkpoint_type: "classification"
+
+  batch_size: <replace_batch_size>
+  num_steps: <replace_train_steps>
+
+  data_augmentation_options {
+    random_horizontal_flip {
+    }
+  }
```
Comment (kevin3-black) [inline rmltraintfbbox/rmltraintfbbox/utils/pipeline_templates/center_net_resnet101_v1_fpn_adam.config:49]: I like to add `random_vertical_flip` and `random_rotation90` to all the pipeline configs as well, since our data is orientation-independent and more augmentations can never hurt

File: rmltraintfbbox/rmltraintfbbox/validation/stats.py [hunk 2]
```diff
@@ -53,7 +54,7 @@ def load_from_dump(cls, dump_path):
         self.stats = {}
         return self
 
-    def add_single_result(self, output, inference_time, bbox, centroid, image_size=None, distance=None):
+    def add_single_result(self, output, true_shape, inference_time, bbox, centroid, image_size=None, distance=None):
```
Comment (kevin3-black) [inline rmltraintfbbox/rmltraintfbbox/validation/stats.py:57]: Could you add a docstring comment for `true_shape`?

File: rmltraintfbbox/setup.py [hunk 3]
```diff
@@ -13,7 +13,7 @@
 
 # determine GPU or CPU install via env variable
 gpu = os.getenv('RML_GPU')
-tensorflow_pkg = 'tensorflow==1.14.0' if not gpu else 'tensorflow-gpu==1.14.0'
+tensorflow_pkg = 'tensorflow==2.3.0' if not gpu else 'tensorflow-gpu==2.3.0'
```
Comment (sshh12) [inline rmltraintfbbox/setup.py:16]: Is this still needed? Iirc tensorflow >=2 will include gpu stuff.

File: rmltraintfbbox/rmltraintfbbox/core.py [hunk 4]
```diff
@@ -200,21 +260,100 @@ def train(ctx: click.Context, train: TrainInput):
...
     return result
     
 
 ### HELPERS ###
+
+def load_fine_tune_checkpoint(
+    model, checkpoint_path, checkpoint_type, checkpoint_version, input_dataset,
+    unpad_groundtruth_tensors):
+
+#NOTE: Moved this function from OD API to add 'expect_partial' to the restore
```
[the remark this answers is not in the document]
Comment (dlfonseka) [reply rmltraintfbbox/rmltraintfbbox/core.py:293]: In the version of this function found in the OD API, the restore tries to "assert_existing_objects_matched" `ckpt.restore(checkpoint_path).assert_existing_objects_matched()`. For some reason, this assertion seems to throw an error. I had to modify the function to remove this assertion and changed the line to `ckpt.restore(checkpoint_path).expect_partial()`, which throws no error. From searching online it seemed like others had run into this problem too - this was my solution.

[the remark this answers is not in the document]
Comment (dlfonseka) [reply rmltraintfbbox/rmltraintfbbox/core.py:293]: Also, forgot to add that this assertion only fails for Centernet. I believe that this is because there is still some code related to Centernet missing from the OD API that would allow Centernet models to be restored from detection checkpoints, and not from classification checkpoints. So it is likely that in the future we can remove this and use the version of the function found in OD API.

File: rmltraintfbbox/setup.py (hunk 3 above)
Comment (dlfonseka) [reply rmltraintfbbox/setup.py]: Yup good catch. Removed gpu version.

File: rmltraintfbbox/rmltraintfbbox/validation/stats.py (hunk 2 above)
Comment (dlfonseka) [reply rmltraintfbbox/rmltraintfbbox/validation/stats.py:57]: Done, thanks for the catch.

File: rmltraintfbbox/rmltraintfbbox/utils/pipeline_templates/center_net_resnet101_v1_fpn_adam.config (hunk 1 above)
Comment (dlfonseka) [reply rmltraintfbbox/rmltraintfbbox/utils/pipeline_templates/center_net_resnet101_v1_fpn_adam.config:49]: Added, thanks for suggestion

File: rmltraintfbbox/rmltraintfbbox/utils/helpers.py [hunk 5]
```diff
@@ -50,11 +50,11 @@ def prepare_for_training(
     hp_metadata = {}
     
     # create a data folder within our base_directory
-    os.makedirs(base_dir / 'data')
+    #os.makedirs(base_dir / 'data')
```
Comment (cSchubes) [inline rmltraintfbbox/rmltraintfbbox/utils/helpers.py:53]: nit: remove

File: rmltraintfbbox/rmltraintfbbox/validation/stats.py [hunk 6]
```diff
@@ -53,11 +54,12 @@ def load_from_dump(cls, dump_path):
...
         """
         Add single inference result to the evaluation.
-        :param output: the parsed output from a BoundingBoxModel inference.
-        :param inference_time: the inference time from a BoundingBoxModel inference.
+        :param output: the parsed output from a detection model inference.
```
Comment (cSchubes) [inline rmltraintfbbox/rmltraintfbbox/validation/stats.py:60]: we use Google style docstrings for everything in the codebase (https://google.github.io/styleguide/pyguide.html#s3.8-comments-and-docstrings), and Sphinx supports them via Napoleon (https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html). would recommend switching back to google style for consistency

File: rmltraintfbbox/rmltraintfbbox/validation/stats.py [hunk 7]
```diff
@@ -88,6 +91,38 @@ def add_single_result(self, output, inference_time, bbox, centroid, image_size=N
         self.centroids.append(centroid)
         self.count += 1
 
+    def parse_inference_output(self, output, image_size):
+        """
+        Parses the raw output of the object detection model into a more sensible format.
+        :param category_index: a category index created with `get_category_index`
```
Comment (cSchubes) [inline rmltraintfbbox/rmltraintfbbox/validation/stats.py:97]: same as above, recommend google style

File: rmltraintfbboxlegacy/README.md [hunk 8]
```diff
@@ -0,0 +1,5 @@
+NOTE: Setup.py pulls an environment variable called RML_GPU to determine
```
Comment (cSchubes) [inline rmltraintfbboxlegacy/README.md:1]: lets remove this note since the GPU flag is no longer used on this plugin

**Review by cSchubes — CHANGES_REQUESTED** [review]
Comments I made are not huge priorities but marking as request changes because of the file permissions changes. Every file is marked "100644 -> 100755". This means every file is now executable, which doesn't make sense for most. It also means files like `rmltraintfbbox/rmltraintfbbox/__init__.py` are marked changed as part of this PR when they actually aren't.

I believe doing a `sudo find /path/to/directory -type d -exec chmod 644 {} \` will fix this. It's necessary to use `find` to avoid changing the permissions of subfolders to 644, which would prevent non-owners from entering them (folders are usually 755).

File: rmltraintfbbox/rmltraintfbbox/core.py [hunk 9]
```diff
@@ -119,78 +122,135 @@ def train(ctx: click.Context, train: TrainInput):
...
+        #main training loop
+        for step in range(1, num_train_steps+1):
+
+            loss = train_step(detection_model, train_input_iterator, optimizer, learning_rate_fn, global_step)
+            
+            if step % 100 == 0:
+                print(f'Training loss at step {step}: {loss}')
+                manager.save()
+
+            if step % 500 == 0:
```
[the remark this answers is not in the document]
Comment (kevin3-black) [reply rmltraintfbbox/rmltraintfbbox/core.py:207]: Looking at a full example run (https://www.comet.ml/seeker-rd/bounding-box/bc7a6607447a4d2fa600c36464366845?experiment-tab=chart&showOutliers=true&smoothing=0&transformY=smoothing&xAxis=step), it seems like all of the train metrics are getting logged properly. I'm guessing this is because comet automatically injects itself into any Tensorboard logging. However, I just noticed that it's actually the eval metrics that are missing, which are arguably more important -- @dlfonseka any idea why?

[the remark this answers is not in the document]
Comment (kevin3-black) [reply rmltraintfbbox/rmltraintfbbox/core.py:207]: Ah, actually, ignore the above. I realized that the metrics being logged are actually the eval metrics, comet just appends `train_` because it is still in a train context. @cSchubes is correct, we should figure out how to log the train loss as well. You should also probably make sure that comet is in the correct context when logging metrics to avoid confusion.

[the remark this answers is not in the document]
Comment (kevin3-black) [reply rmltraintfbbox/rmltraintfbbox/core.py:207]: On the first point, I don't think that the frequency of logging/evaluation/checkpoint saving should be based on the total number of train steps -- every fixed amount of steps is the right way to do it. However, it would be nice if these numbers were configurable, since 100/500 might be too frequent or infrequent depending on how fast your hardware is.

File: rmltraintfbbox/rmltraintfbbox/validation/stats.py (hunk 6 above)
Comment (kevin3-black) [reply rmltraintfbbox/rmltraintfbbox/validation/stats.py:60]: This is my bad, I wrote this code and I'm used to reST style. I used it for all of starfish, and a bunch of docstrings in all the train plugins I wrote as well. I can go back and change them all later, although I don't think we should block this PR on that.

File: rmltraintfbbox/rmltraintfbbox/core.py [hunk 10]
```diff
@@ -119,78 +122,135 @@ def train(ctx: click.Context, train: TrainInput):
...
-    num_train_steps = int(num_train_steps)
-
-    tf_config = tf.estimator.RunConfig(model_dir=model_dir)
-    train_and_eval_dict = model_lib.create_estimator_and_inputs(
-        run_config=tf_config,
-        sample_1_of_n_eval_examples=1,
-        hparams=model_hparams.create_hparams(None),
-        pipeline_config_path=pipeline_config_path,
-        train_steps=num_train_steps)
+    num_train_steps = int(metadata['hyperparameters']['train_steps'])
```
[the remark this answers is not in the document]
Comment (dlfonseka) [reply rmltraintfbbox/rmltraintfbbox/core.py:125]: Does it matter that we cast to int if an int isn't passed in?

File: rmltraintfbbox/rmltraintfbbox/core.py [hunk 11]
```diff
@@ -200,21 +260,100 @@ def train(ctx: click.Context, train: TrainInput):
...
+    #zip files in export directory and add to extra_files
+    #shutil.make_archive(os.path.join(saved_model_dir,'export'), 'zip', model_dir, saved_model_dir)
+    shutil.make_archive(os.path.join(saved_model_dir,'export'), 'zip', model_dir, 'export')
+    extra_files.append(os.path.join(saved_model_dir, 'export.zip'))
+    saved_model_path = os.path.join(saved_model_dir, 'saved_model', 'saved_model.pb' )
+    
     if comet:
         experiment.log_asset_data(train.metadata, file_name="metadata.json")
 
     # export metadata locally
```
[the remark this answers is not in the document]
Comment (dlfonseka) [reply rmltraintfbbox/rmltraintfbbox/core.py:292]: The issue is that the upload function in raven core expects a .pb file, but our exported model requires a whole directory of stuff to restore from. Have talked to Abhi about allowing us to pass in our zip file as the exported model - then I think we wouldn't have to upload the zip folder as an extra file.

File: rmltraintfbbox/rmltraintfbbox/core.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (dlfonseka) [reply rmltraintfbbox/rmltraintfbbox/core.py]: Steps between train/eval logging to comet now configurable in the plugin specific options of the config files. The naming scheme of the comet logging is also fixed.

File: rmltraintfbbox/rmltraintfbbox/core.py [hunk 12]
```diff
@@ -55,20 +62,14 @@ def train(ctx: click.Context, train: TrainInput):
...
     config = train.plugin_config
     metadata = train.plugin_metadata
     comet = config.get('comet')
 
-    # set up TF verbosity
-    if config['verbose']:
-        tf.compat.v1.logging.set_verbosity(tf.compat.v1.logging.INFO)
-    else:
-        tf.compat.v1.logging.set_verbosity(tf.compat.v1.logging.FATAL)
+    if not config['verbose']:
```
[the remark this answers is not in the document]
Comment (dlfonseka) [reply rmltraintfbbox/rmltraintfbbox/core.py]: Changed verbosity to match tf2 procedure

File: rmltraintfbbox/rmltraintfbbox/core.py [hunk 13]
```diff
@@ -119,78 +125,145 @@ def train(ctx: click.Context, train: TrainInput):
...
+                if comet:
+                    experiment.log_metric('loss', loss)
+                
+            if step % config.get('log_eval_every') == 0:
+                stack.enter_context(experiment.validate())
+                manager.save()
+                eval_metrics = evaluate(detection_model, configs, eval_input, global_step)
+                if comet:
+                    experiment.log_metrics(eval_metrics, step=step)
+                stack.enter_context(experiment.train())
```
Comment (kevin3-black) [inline rmltraintfbbox/rmltraintfbbox/core.py:217]: Anything with `experiment` should go inside an `if comet:` block. Also, should probably add averaging for the train loss like we discussed.

File: rmltraintfbbox/rmltraintfbbox/core.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (cSchubes) [reply rmltraintfbbox/rmltraintfbbox/core.py:128]: it would if what got passed in couldn't be cast as an int - since we don't have any field validation right now, the `metadata['hyperparameters']['train_steps']` could be an array, or string, or anything else

not a big deal for now though. it'll just fail without good warning to the user

File: rmltraintfbbox/rmltraintfbbox/core.py [hunk 14]
```diff
@@ -119,78 +122,135 @@ def train(ctx: click.Context, train: TrainInput):
...
+                if comet:
+                    experiment.log_metrics(metrics, step=step)
+                
+
+        training_time = time.time() - start
+        
+        click.echo(f'Training complete. Took {training_time} seconds.')
 
     # final metadata and return of TrainOutput object
     metadata['date_completed_at'] = datetime.utcnow().isoformat() + "Z"
```
[the remark this answers is not in the document]
Comment (cSchubes) [reply rmltraintfbbox/rmltraintfbbox/core.py]: added to backlog for raven core, resolving

File: rmltraintfbbox/rmltraintfbbox/validation/stats.py (hunk 6 above)
Comment (cSchubes) [reply rmltraintfbbox/rmltraintfbbox/validation/stats.py:60]: I've never looked at the docstrings in those before, didn't realize. Not a huge deal, I just find the Google style ones easier to read. Resolving for now

File: rmltraintfbbox/rmltraintfbbox/utils/helpers.py (hunk 5 above)
Comment (cSchubes) [reply rmltraintfbbox/rmltraintfbbox/utils/helpers.py:53]: leaving as a nit

File: rmltraintfbboxlegacy/README.md (hunk 8 above)
Comment (cSchubes) [reply rmltraintfbboxlegacy/README.md:1]: leaving as nit

**Review by cSchubes — CHANGES_REQUESTED** [review]
- All files in the new `rmltraintfbboxlegacy/rmltraintfbboxlegacy/utils/checkpoint_0/` folder are still changed to permission level `755` rather than the original and default `644`.
- All files in the new `rmltraintfbboxlegacy/rmltraintfbboxlegacy/utils/model_defaults/` folder are still changed to permission level `755` rather than the original and default `644`.
All files in the new `rmltraintfbboxlegacy/rmltraintfbboxlegacy/utils/pipeline_templates/` folder are still changed to permission level `755` rather than the original and default `644`.

I'd recommend just doing the same process I recommended before for the `rmltraintfbbox` directory on the `rmltraintfbboxlegacy` directory. Sorry for such nitpicking - just want to keep the git history clean and not add a bunch of pseudochanges that aren't "real" or desired.

Other than that, looks great. Let's get this merged in!

**Review by cSchubes — APPROVED** [review]
